Objects/frameobject.c

```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
                 case POP_JUMP_IF_FALSE:
                 case POP_JUMP_IF_TRUE:
                 case JUMP_IF_NOT_EXC_MATCH:
-                    j = get_arg(code, i);
+                {
+                    int j = get_arg(code, i);
                     assert(j < len);
-                    if (blocks[j] == -1 && j < i) {
+                    if (blocks[j] == UNINITIALIZED && j < i) {
                         todo = 1;
                     }
-                    assert(blocks[j] == -1 || blocks[j] == block_stack);
-                    blocks[j] = block_stack;
+                    if (opcode == JUMP_IF_NOT_EXC_MATCH) {
+                        block_stack = pop_block(block_stack);
+                    }
+                    int64_t target_stack = block_stack;
+                    if (opcode != JUMP_IF_FALSE_OR_POP && opcode != JUMP_IF_TRUE_OR_POP) {
+                        target_stack = pop_block(target_stack);
+                    }
+                    block_stack = pop_block(block_stack);
+                    assert(blocks[j] == UNINITIALIZED || blocks[j] == target_stack);
+                    blocks[j] = target_stack;
                     blocks[i+1] = block_stack;
                     break;
+                }
                 case JUMP_ABSOLUTE:
                     j = get_arg(code, i);
                     assert(j < len);
-                    if (blocks[j] == -1 && j < i) {
+                    if (blocks[j] == UNINITIALIZED && j < i) {
                         todo = 1;
                     }
-                    assert(blocks[j] == -1 || blocks[j] == block_stack);
+                    assert(blocks[j] == UNINITIALIZED || blocks[j] == block_stack);
                     blocks[j] = block_stack;
                     break;
-                case SETUP_FINALLY:
-                    j = get_arg(code, i) + i + 1;
-                    assert(j < len);
-                    except_stack = push_block(block_stack, Except);
-                    assert(blocks[j] == -1 || blocks[j] == except_stack);
-                    blocks[j] = except_stack;
-                    block_stack = push_block(block_stack, Try);
-                    blocks[i+1] = block_stack;
-                    break;
-                case SETUP_WITH:
-                case SETUP_ASYNC_WITH:
-                    j = get_arg(code, i) + i + 1;
-                    assert(j < len);
-                    except_stack = push_block(block_stack, Except);
-                    assert(blocks[j] == -1 || blocks[j] == except_stack);
-                    blocks[j] = except_stack;
-                    block_stack = push_block(block_stack, With);
+                case POP_EXCEPT:
+                    block_stack = pop_block(pop_block(pop_block(block_stack)));
```

Comment:
Because that's what POP_EXCEPT does; it pops three values from the *evaluation* stack.
I suspect that would have clearer if I'd renamed `block_stack` :slightly_smiling_face:

Reply:
Yeah, changing the names from "block" to "stack" helps a lot.  Thanks!